doc/mpls/MPLS_hld.md

```diff
@@ -0,0 +1,510 @@
+
+### High-Level Design
+
+#### Overview
+![Overview diagram](images/MPLS_overview_diagram.png "Overview of MPLS components")
+
+**Figure 1: Overview of the data flow and related components of MPLS**
+
+#### Database Changes
+This section describes the modifications to SONiC Databases to support MPLS.
+
+##### APPL_DB
+
+###### INTERFACE_TABLE
+The existing INTERFACE_TABLE is enhanced to accept a new "mpls" enable/disable attribute.
+
+``` rfc5234
+INTERFACE_TABLE|{{interface_name}}
+    "mpls":{{enable|disable}} (OPTIONAL)
+
+; Defines schema for MPLS configuration attribute
+key                         = INTERFACE:ifname             ; Interface name
+; field                     = value
+mpls                        = "enable" / "disable"         ; Enable/disable MPLS function. Default disable
+```
+
+###### ROUTE_TABLE
+The existing ROUTE_TABLE for IPv4/IPv6 prefix routes is enhanced to accept an optional MPLS label stack component in the existing "nexthop" formatted-string attribute.  The new format of the nexthop attribute string with MPLS label stack is: "label0/.../labelN+ip-prefix".  For IP-only next hops, the previous string format is retained and no preceding label stack information will be included.
+
+``` rfc5234
+"ROUTE_TABLE":{{prefix}}
+    "nexthop":{{nexthop_list}}
+    "ifname":{{ifname_list}}
+
+; Defines schema for IPv4/IPv6 route table attributes
+key                         = ROUTE_TABLE:prefix       ; IPv4/IPv6 prefix
+; field                     = value
+nexthop                     = nexthop_list             ; List of nexthops.
+ifname                      = ifname_list              ; List of interfaces.
+;value annotations
+ifname                      = 1*64VCHAR                ; name of the Interface (Port Channel)
+```
+
+###### LABEL_ROUTE_TABLE
+A new LABEL_ROUTE_TABLE is introduced for MPLS in-segment entries to accept the same attributes as ROUTE_TABLE:
+- A "nexthop" formatted-string attribute with optional MPLS label stack component.
+- A "ifname" attribute.
+
+``` rfc5234
+"LABEL_ROUTE_TABLE":{{mpls_label}}
+    "nexthop":{{nexthop_list}}
+    "ifname":{{ifname_list}}
+
+; Defines schema for MPLS label route table attributes
+key                         = LABEL_ROUTE_TABLE:mpls_label ; MPLS label
+; field                     = value
+nexthop                     = STRING                   ; Comma-separated list of nexthops.
+ifname                      = STRING                   ; Comma-separated list of interfaces.
+```
+
+#### Software Modules
+This section describes modifications to SONiC infrastructure software modules to support MPLS.
+
+##### NetLink
+Netlink is an existing open source library imported to SONiC.
+Modifications to the existing NetLink MPLS implementation were needed to support MPLS attributes.
+###### Functions
+New accessors were added to retrieve the MPLS NH destination and TTL values:
+```
```

Comment:
I assume the kernel is getting updated with MPLS label for any origination control traffic to send with MPLS labels from FRR, please clarify.

Reply:
Kernel will be getting updated from cRPD installed MPLS routes.